044-Stable-identities-for-Kafka-Connect-worker-nodes.md

```diff
@@ -0,0 +1,176 @@
+Each connector task is running on a specific worker node with given identity (IP address).
+
+When changes need to be rolled out to the Kafka Connect cluster, Strimzi CO will modify the Kubernetes Deployment.
+Based on the changes, Kubernetes will roll the Connect Pods.
+By default, Strimzi configures the Deployment to use the `RollingUpdate` strategy.
+So out of the box, Kubernetes will:
+1) Start a new Kafka Connect Pod
+2) Wait until the new Pod gets ready
+3) Stop the old Connect Pod
+4) Move to next Pod (if any) and repeat the same
+
+The new Pod will be assigned a new IP address which it will use for its advertised address.
+Because of the new address, it will have a new _identity_.
+So during the rolling update, the Kafka Connect cluster will not see this operation as a restart of existing worker node.
+Instead, it will see it as a new worker node joining the cluster and later a previously existing node leaving the cluster.
+
+_Note: Optionally, users can enable the `Recreate` deployment strategy in the `KafkaConnect` custom resource._
+_In that case, Kubernetes will first delete the old Pod and only then create the new Pod._
+_However, the new Pod still gets a new IP address so the identity of the Pod changes._
+
+This means that the worker nodes where the connector tasks are scheduled all disappear during the rolling update and the tasks need to be rescheduled.
+It will take some time (by default 5 minutes) for the cluster to detect that the old nodes left the cluster and reschedule the tasks on other nodes.
+Until this happens, it is possible that the tasks will be scheduled to one of the pods which was deleted and are thus not running.
+
+## Proposal
+
+Giving the Kafka Connect workers stable identities should improve the scheduling of the connector tasks and minimize the number of task rebalances.
+Since the worker identity is based on the Pod address, it means that we need to give the pods stable addresses.
+
+To achieve that, we should migrate the Kafka Connect clusters from Kubernetes Deployment resource to the StrimziPodSet resources.
+The Pods created by the StrimziPodSet will use stable pod names instead of the randomized names used for the Pods managed by Kubernetes Deployments.
+The pods will be named `<cluster-name>-connect-<index>` where the `<cluster-name>` is the name of the custom resource and the `<index>` will be a sequential index number starting with 0.
+For example, the pod names for Kafka Connect cluster created by `KafkaConnect` custom resource named `my-connect` with 3 replicas will be `my-connect-connect-0`, `my-connect-connect-1`, and `my-connect-connect-2`
+This pattern follows the way the pods are currently named as well.
+
+In addition to the StrimziPodSet resource used to manage the Pods, Strimzi will create a headless Kubernetes Service to give these pods a stable DNS names.
+The headless service will be named `<cluster-name>-connect`.
```

Comment:
`<cluster-name>-worker`?

Reply:
This follows the existing pattern. I think we should keep it to minimize the disruption on users and their possible monitoring tools etc.